client/src/query.rs

```diff
@@ -0,0 +1,310 @@
+    Single(#[from] SingleQueryError),
+    /// Other error
+    Other(#[from] eyre::Error),
+}
+
+impl From<ResponseReport> for ClientQueryError {
+    #[inline]
+    fn from(ResponseReport(err): ResponseReport) -> Self {
+        Self::Other(err)
+    }
+}
+
+impl QueryExecutor for Client {
+    type Cursor = ClientQueryCursor;
+    type Error = ClientQueryError;
+    type SingleError = ClientQueryError;
+
+    fn execute_singular_query(
+        &self,
+        query: SingularQueryBox,
+    ) -> Result<SingularQueryOutputBox, Self::Error> {
+        let request_head = self.get_query_request_head();
+
+        let request = QueryRequest::Singular(query);
+
+        let response = request_head.assemble(request).build()?.send()?;
+        let response = decode_singular_query_response(&response)?;
+
+        Ok(response)
+    }
+
+    fn start_query(
+        &self,
+        query: IterableQueryWithParams,
+    ) -> Result<(IterableQueryOutputBatchBox, Option<Self::Cursor>), Self::Error> {
+        let request_head = self.get_query_request_head();
+
+        let request = QueryRequest::StartIterable(query);
+
+        let response = request_head.assemble(request).build()?.send()?;
+        let response = decode_iterable_query_response(&response)?;
+
+        let (batch, cursor) = response.into_parts();
+
+        let cursor = cursor.map(|cursor| ClientQueryCursor {
+            request_head,
+            cursor,
+        });
+
+        Ok((batch, cursor))
+    }
+
+    fn continue_query(
+        cursor: Self::Cursor,
+    ) -> Result<(IterableQueryOutputBatchBox, Option<Self::Cursor>), Self::Error> {
+        let ClientQueryCursor {
+            request_head,
+            cursor,
+        } = cursor;
+
+        let request = QueryRequest::ContinueIterable(cursor);
+
+        let response = request_head.assemble(request).build()?.send()?;
+        let response = decode_iterable_query_response(&response)?;
+
+        let (batch, cursor) = response.into_parts();
+
+        let cursor = cursor.map(|cursor| ClientQueryCursor {
+            request_head,
+            cursor,
+        });
+
+        Ok((batch, cursor))
+    }
+}
+
+impl Client {
+    /// Get a [`ClientQueryRequestHead`] - an object that can be used to make queries independently of the client.
+    ///
+    /// You probably do not want to use it directly, but rather use [`Client::query_single`] or [`Client::query`].
+    fn get_query_request_head(&self) -> ClientQueryRequestHead {
+        ClientQueryRequestHead {
+            torii_url: self.torii_url.clone(),
+            headers: self.headers.clone(),
+            account_id: self.account.clone(),
+            key_pair: self.key_pair.clone(),
+        }
+    }
+
+    /// Execute a singular query and return the result
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the query execution fails.
+    pub fn query_single<Q>(&self, query: Q) -> Result<Q::Output, ClientQueryError>
```

Comment:
I feel like whether or not we implement projections, this will be a wasted effort.

If we implement projections, we are likely to be able to get rid of all singular queries (#4933), making the `FindExecutorDataModel` a special case. With no singular queries - no need for singular builders.

If we don't implement projections then I don't see any options we can define for the singular queries..

Reply:
let's close this